### ISSUE TITLE
Windows Defender Flags Runtime Installers as Trojans

### ISSUE THREAD

**Issue by jbhurst** [KhronosGroup/Vulkan-LoaderAndValidationLayers#1640]
This seems related to the previus case: https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/issues/474

This morning, 3/31/2017, Windows Defender on Windows 10 x64 Pro reported the following:
"
The following error occurred: Error code 0x80070057. The parameter is incorrect. 

Category: Trojan

Description: This program is dangerous and executes commands from an attacker.

Recommended action: Remove this software immediately.

Items: 
file:C:\Program Files (x86)\VulkanRT\1.0.26.0\UninstallVulkanRT.exe
file:C:\Program Files (x86)\VulkanRT\1.0.26.0\V.ico
regkey:HKLM\SOFTWARE\MICROSOFT\WINDOWS\CURRENTVERSION\UNINSTALL\VulkanRT1.0.26.0
regkeyvalue:HKLM\SOFTWARE\Microsoft\Windows Defender\Exclusions\Paths\\G:\
regkeyvalue:HKLM\SOFTWARE\Microsoft\Windows Defender\Exclusions\Paths\\J:\
uninstall:HKLM\SOFTWARE\MICROSOFT\WINDOWS\CURRENTVERSION\UNINSTALL\VulkanRT1.0.26.0

Get more information about this item online.
"

My virus and spyware definition versions are 1.239.494.0.

I don't know if this is an MS issue or a Vulkan one, but it looked like last time, Khronos was able to do something to resolve.

**Comment by lenny-lunarg** [comment]
In general, we do try to prevent our installers and uninstallers from being flagged by antiviruses, but it's essentially impossible to avoid it entirely. In this case, you're using a fairly old runtime (I think we released 1.0.26.0 sometime around July or August). There's really nothing we can do to prevent old installers from being flagged, since we don't change them. All we can do is release newer installers where we attempt to resolve the issue.

You said you're not sure if it's an MS issue or a Vulkan one. It's really an MS issue, since it's their software that's falsely reporting a virus, but like you said we do have some tricks to change around what the software reports (essentially we just recompile with different options to cause the code to look different).

Since you're reporting this against an old runtime, there's not really anything I can do except report the false positive to MS (and they would likely ignore me). If this issue exists for the newest (1.0.42.2) installer, then it's worth making sure we avoid the problem in our next SDK, but I would be surprised if Windows Defender was flagging the newest runtime without us having heard about it.

Since we can't really do anything about antiviruses flagging old runtimes, I'm going to close this issue. My advice is to update your runtime, either by updating your graphics drivers or by installing an SDK (you can feel free to manually uninstall the 1.0.26.0 runtime if your system is unhappy with it). If you find that your system flags the 1.0.42.2 files, then please do reopen the issue, but there really isn't anything we can do about old installers, except report them to the antivirus vendor.

**Comment by jbhurst** [comment]
Thanks for the info.  I didn't realize I had an old runtime.  I'll get the new driver.

**Comment by ironiridis** [comment]
For anyone else trying to search for this issue, I am receiving similar warnings from Windows Defender. It's being classified as:
`Trojan:Win32/Suweezy`
`Trojan:Win32/Xadupi`

It appears to be in version 1.0.3.1 and is matching all of the files in the directory, including `LICENSE.txt`, `VULKANRT_LICENCE.rtf`, and `V.ico`. This suggests to me that this isn't merely a false positive, but that the malware is actually using the Vulkan runtime and Microsoft has incorrectly flagged Vulkan because it arrived in a sample.

**Comment by lenny-lunarg** [comment]
The way I read the error message from @jbhurst, I don't think it was finding anything wrong with the files that the runtime installed. It looks to me like it simply believes you should remove them because they were created by the installer and it thinks the installer is a virus. It also says that the registry keys that the runtime created should be deleted, and the registries are in integral part of Windows (so not a virus). So I think it's just saying you should revert everything that the installer did. I can't say if that's the same for @ironiridis, since I don't have the error message, but my guess would be that it's the same.

Also, 1.0.3.1 is more than a year old, and was back before we had taken any measures to avoid antivirus false positives. So it doesn't surprise me that Windows would think that one is a virus. We've sent runtimes to antivirus vendors before, and every response we've received has indicated that it was a false positive. I can't remember if we've had 1.0.3.1 specifically tested, but I'd be very surprised if that was a virus. 

On top of that, running the 1.0.3.1 runtime installer through VirusTotal, which aggregates the results of a number of different antiviruses says that [0 / 62](https://www.virustotal.com/en/file/32be60a2ae99f7a68ed7be949ae9cbe2158b4784018f72e4e3cf1e26923ade13/analysis/) antiviruses flag it as a virus. This makes me feel strongly that Windows Defender is simply reporting a false positive.

**Comment by Maik268** [comment]
I have the same warning with Windows Defender now.
I'm using VulkanRT 1.0.39.1 which is included in latest nvidia driver.
![unbenannt](https://cloud.githubusercontent.com/assets/13015934/24567750/5490eac2-165f-11e7-83c1-95f4f3674bba.PNG)

**Comment by lenny-lunarg** [comment]
If Windows Defender is flagging 1.0.39.1 as well, then we probably have a problem. I just ran Windows Defender on the runtime installers for 1.0.3.1, 1.0.39.1, and 1.0.42.2 and Windows Defender didn't flag any of them. I'm using virus definition version 1.239.506.0 (same as @ironiridis). Enough people have reported this issue that I'm quite sure this is a real issue &mdash; my computer must just be different in some way.

I'm using Windows 10 Home, Version 1607, OS Build 14393.953. If any of you could report the exact version of Windows and of your virus definitions, that would be helpful. Ideally, I'd like to figure out a way to get Windows Defender to mark a runtime installer as a virus on a machine I have access to. If I can't do that, then all I'll be able to do is file a false positive report with Microsoft (which I'll probably do regardless).

**Comment by kribjo** [comment]
Windows Defender catches 1.0.42.2 for me too:

"The following error occurred: Error code 0x80070057. The parameter is incorrect. 

Category: Trojan

Description: This program is dangerous and executes commands from an attacker.

Recommended action: Remove this software immediately.

Items: 
file:C:\Program Files (x86)\vulkanrt\1.0.42.2\LICENSE.txt
file:C:\Program Files (x86)\vulkanrt\1.0.42.2\UninstallVulkanRT.exe
file:C:\Program Files (x86)\vulkanrt\1.0.42.2\V.ico
file:C:\Program Files (x86)\vulkanrt\1.0.42.2\vulkaninfo.exe
file:C:\Program Files (x86)\vulkanrt\1.0.42.2\vulkaninfo32.exe
file:C:\Program Files (x86)\vulkanrt\1.0.42.2\VULKANRT_LICENSE.rtf
folder:C:\Program Files (x86)\vulkanrt\
folder:C:\Program Files (x86)\vulkanrt\1.0.42.2\
regkey:HKLM\SOFTWARE\MICROSOFT\WINDOWS\CURRENTVERSION\UNINSTALL\VulkanRT1.0.42.2
regkeyvalue:HKLM\SOFTWARE\Microsoft\Windows Defender\Exclusions\Paths\\C:\Program Files (x86)\Intel\Bluetooth\
uninstall:HKLM\SOFTWARE\MICROSOFT\WINDOWS\CURRENTVERSION\UNINSTALL\VulkanRT1.0.42.2"

Identifies it as Trojan:Win32/Xadupi and Trojan:Win32/Suweezy
https://www.microsoft.com/security/portal/threat/encyclopedia/entry.aspx?name=Trojan%3aWin32%2fXadupi&threatid=2147709752&enterprise=0
https://www.microsoft.com/security/portal/threat/encyclopedia/entry.aspx?name=Trojan%3aWin32%2fSuweezy&threatid=2147716113&enterprise=0

**Comment by kribjo** [comment]
VirusTotal report for 1.0.39.1 that was included in latest Nvidia drivers:
https://www.virustotal.com/en/file/1f30c3ec401b2af80a528550020667743a60b2fc8d59621f3774ff2a1a8685b7/analysis/

SHA256:	1f30c3ec401b2af80a528550020667743a60b2fc8d59621f3774ff2a1a8685b7
File name:	VulkanRT-Installer.exe
Detection ratio:	2 / 61
Analysis date:	2017-03-31 20:21:33 UTC ( 35 minutes ago )

**Comment by kribjo** [comment]
Just got new definitions for Windows Defender, version 1.239.517.0, still same detection of v1.0.42.2 files.

**Comment by Maik268** [comment]
OS: Windows 10 Pro, Version 1607 (Build 14393.969)
Defender: Antivirus: 1.239.517.0, Spyware: 1.239.517.0
Vulkan: 1.0.39.1

But scanning the folder now manually does not show any results?

**Comment by lenny-lunarg** [comment]
I just submitted a false positive report to Microsoft's support center with the 1.0.42.2 runtime installer. Hopefully they'll be able to do something.

It looks like I'm running very nearly the same OS as @Maik268, which unfortunately really doesn't clear anything up. I'm running a full system scan now with 1.0.3.1, 1.0.39.1, and 1.0.42.2 all installed. Hopefully that will flag something, but we'll see. I'll post anything new that comes up here.

**Comment by lenny-lunarg** [comment]
I actually got a very quick response from Microsoft. I've pasted the worthwhile part below:

```
The Microsoft Malware Protection Center (MMPC) has investigated the following file(s) which we received on 3/31/2017 2:09:39 PM Pacific Time.
Below is the determination for your submission.

========
Submission ID MMPC17033131569576

  Submitted Files
  =============================================
  VulkanRT-1.0.42.2-Installer.zip [Not Malware]
  +---VulkanRT-1.0.42.2-Installer.exe [Not Malware]


Analyst Comments
================
-- 3/31/2017 2:27:04 PM --
This has been identified as an FP (false positive) and these false detections have been fixed and will be released in VDM build 1.239.530.0 and above.
SDN has been issued for the signatures and detections should not happen anymore for cloud-connected clients.




Your submission was scanned using antimalware definition version 1.239.517.0.
========
```

I have no idea when this will go live, but it sounds like it should be fixed in the near future. Whenever this does get released I'll check if it only applies to 1.0.42.2. If that's the case I'll just submit all of the runtime installers to try to get Windows Defender to stop flagging them. I'm going to close this issue now. If 1.239.530.0 releases and the problem isn't fixed, then we'll reopen the issue.

**Comment by kribjo** [comment]
I updated to 1.239.530.0 definitions and they seem to play nicely with both 1.0.39.1 and 1.0.42.2.
